src/modules/data/TxInput.ts

```diff
@@ -0,0 +1,129 @@
+/*******************************************************************************
+
+    The class that defines the transaction's inputs of a block.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+import { Hash, hashFull, makeUTXOKey } from './Hash';
+import { Signature } from './Signature';
+import { Transaction } from './Transaction';
+import { Validator, ITxInput } from './validator';
+
+import { SmartBuffer } from 'smart-buffer';
+
+/**
+ * The class that defines the transaction's inputs of a block.
+ * Convert JSON object to TypeScript's instance.
+ * An exception occurs if the required property is not present.
+ */
+export class TxInput
+{
+    /**
+     * The hash of the UTXO to be spent
+     */
+    public utxo: Hash;
+
+    /**
+     * A signature that should be verified using public key of the output in the previous transaction
+     */
+    public signature: Signature;
+
+    /**
+     * Constructor
+     * @param first  UTXO or hash of transaction or instance of Transaction
+     * @param second Instance of Signature or index of Outputs
+     */
+    constructor (first: Hash | Transaction, second?: Signature | bigint)
+    {
+        // constructor (first: Hash, second: Signature) {}
+        // first is the hash of the utxo
+        // second is the instance of the Signature
+        if ((first instanceof Hash) && (second instanceof Signature))
+        {
+            this.utxo = first;
+            this.signature = second;
+        }
+        // constructor (first: Transaction, second: bigint) {}
+        // first is the instance of the previous transaction
+        // second is the index of the output in the previous transaction
+        else if ((first instanceof Transaction) && (typeof second === 'bigint'))
+        {
+            this.utxo = makeUTXOKey(hashFull(first), second);
+            this.signature = Signature.init;
+        }
```

Comment:
IMO this jungle of `if` on types is a big red flag. It completely nullify the benefit of having a typed language, which is to catch errors at compile time, not run time.

If we want to write code that uses an overload-like approach, okay, but it should be done using the language feature that allows it: https://www.typescriptlang.org/docs/handbook/functions.html#overloads